doc/ui_component2_transition.md

```diff
@@ -0,0 +1,132 @@
+
+Migration can happen at any time, and is opt-in. We'll provide a [codemod](#updating) to handle the simpler cases and 
+run it as 
+part of cleanup, but it will take manual intervention to update some components.
+
+## Why a new base class?
+- It's not possible for us to automatically support both unsafe and safe lifecycle methods without a resolved AST in 
+the over_react builder, which would increase build times dramatically.
+- We needed some small breakages to support
+  - JS-backed maps
+  - New lifecycle methods
+  - Injection of Dart->JS interop pieces into component instance (initComponentInternal) needed to change (this was 
+ publicly accessible, but is not done anywhere)
+  - componentDidUpdate, which now takes an additional `snapshot` argument
+  - Assignment of `props`/`state` to arbitrary Maps (was not advised, but possible)
+  - Prop validation, which had to be reimplemented since it relied on unsafe component lifecycle transitions
+
+## New Features, improvements, and breakages
+
+### Small Perf Improvements
+Component2 provides a small performance boost over Component, as a result of JS-backed maps and other optimizations.
+
+#### Data
+- For a component that reads a single prop and renders a simple div, upgrading it to UiComponent2 results in:
+  - initial renders that are 11% faster
+  - rerenders that are 6.5% faster
+- Special case: for a component that has simple lifecycles, but heavy prop reads, upgrading it to 
+UiComponent2 results in:
+  - initial renders that are 12% faster
+  - rerenders that are 28% faster
+  
+### JS-Backed Maps
+Props/state key-value pairs are stored directly on JS objects, instead of within Dart Maps which are opaque to React. Results in:
+
+- Modest overall rendering perf improvements due to decreased copying, garbage creation, and Dart map overhead
+- Props/state show up in the React DevTools, and primitive values (strings, numbers, booleans) can be edited live
+- Easier to interop with JS libraries that manipulate props, like react-redux
+- Easier to maintain react-dart/over_react interop and add new features
+
+### Lifecycle Method Changes
+
+- Drops support for unsafe lifecycle methods:
+  - componentWillMount
+  - componentWillReceiveProps
+  - componentWillUpdate
+- Adds support for new, safe lifecycle methods:
+  - getDerivedStateFromProps
+  - getSnapshotBeforeUpdate
+  - componentDidCatch / getDerivedStateFromError (adds support for [error boundaries](https://reactjs.org/docs/error-boundaries.html))
+- componentDidUpdate now receives a `snapshot` argument, which is the return value of getSnapshotBeforeUpdate.
+  
+See the [React docs](https://reactjs.org/docs/react-component.html#the-component-lifecycle) for information on why the old methods are unsafe and how to transition to the new ones. This 
+information is also available and linked to within the lifecycle methods' doc comments.
+
+### ErrorBoundary Component Added
```

Comment:
I feel like this section should go after "Other Method Changes" or before "Lifecycle Method Changes" for continuity.

Reply:
Ah yep, good call 😄